File: python/tests/integration/_async/test_command_environment.py
"""Integration tests for command environment variables and advanced features (cwd, envs, new return format)."""

import os

import pytest
import pytest_asyncio

from agentbay import AsyncAgentBay
from agentbay import CreateSessionParams


@pytest_asyncio.fixture(scope="module")
async def agent_bay():
    api_key = os.environ.get("AGENTBAY_API_KEY")
    if not api_key:
        pytest.skip("AGENTBAY_API_KEY environment variable not set")
    return AsyncAgentBay(api_key=api_key)


@pytest_asyncio.fixture
async def test_session(agent_bay):
    """Create a test session."""
    result = await agent_bay.create(CreateSessionParams(image_id="linux_latest"))
    if not result.success:
        pytest.skip(f"Failed to create session: {result.error_message}")
    yield result.session
    await result.session.delete()


@pytest.mark.asyncio
async def test_command_new_return_format(test_session):
    """Test command execution with new return format (exit_code, stdout, stderr)."""
    cmd = test_session.command
    result = await cmd.execute_command("echo 'Hello, AgentBay!'")
    
    # Verify new fields exist
    assert hasattr(result, 'exit_code'), "exit_code field should exist"
    assert hasattr(result, 'stdout'), "stdout field should exist"
    assert hasattr(result, 'stderr'), "stderr field should exist"
    
    # Verify success case
    assert result.success, "Command should succeed"
    assert result.exit_code == 0, "Exit code should be 0 for success"
    assert "Hello, AgentBay!" in result.stdout, "Stdout should contain expected output"
    assert result.output == result.stdout, "Output should equal stdout for success"
    
    print(f"✓ New return format test passed: exit_code={result.exit_code}, stdout={result.stdout}")


@pytest.mark.asyncio
async def test_command_error_with_exit_code(test_session):
    """Test error command with exit_code, stderr, and trace_id."""
    cmd = test_session.command
    result = await cmd.execute_command("ls /non_existent_directory_12345")
    
    # Verify error case
    assert hasattr(result, 'exit_code'), "exit_code field should exist"
    assert hasattr(result, 'stderr'), "stderr field should exist"
    assert hasattr(result, 'trace_id'), "trace_id field should exist"
    
    # Error commands should have non-zero exit code
    # Note: success field behavior depends on implementation
    if result.exit_code != 0:
        assert result.exit_code != 0, "Exit code should be non-zero for error"
        # trace_id is optional, only present when errorCode != 0
        if result.trace_id:
            print(f"✓ Error command test passed: exit_code={result.exit_code}, stderr={result.stderr}, trace_id={result.trace_id}")
        else:
            print(f"✓ Error command test passed: exit_code={result.exit_code}, stderr={result.stderr} (no trace_id)")
    else:
        # If exit_code is 0, the command might have succeeded in some way
        print(f"⚠ Command returned exit_code=0, but this is acceptable")


@pytest.mark.asyncio
async def test_command_with_cwd(test_session):
    """Test command execution with cwd parameter."""
    cmd = test_session.command
    result = await cmd.execute_command("pwd", cwd="/tmp")
    
    assert result.success, "Command should succeed"
    assert result.exit_code == 0, "Exit code should be 0"
    # The output should contain /tmp or be /tmp
    assert "/tmp" in result.stdout, f"Working directory should be /tmp, got: {result.stdout}"
    
    print(f"✓ CWD test passed: working directory={result.stdout.strip()}")


@pytest.mark.asyncio
async def test_command_with_envs(test_session):
    """Test command execution with envs parameter."""
    cmd = test_session.command
    result = await cmd.execute_command(
        "echo $TEST_VAR",
        envs={"TEST_VAR": "test_value_123"}
    )
    
    assert result.success, "Command should succeed"
    assert result.exit_code == 0, "Exit code should be 0"
    # The environment variable should be set
    # Note: This depends on backend implementation
    output = result.stdout.strip()
    if "test_value_123" in output:
        print(f"✓ Envs test passed: environment variable set correctly: {output}")
    else:
        print(f"⚠ Envs test: environment variable may not be set (output: {output})")
        # This is acceptable if backend doesn't support envs yet


@pytest.mark.asyncio
async def test_command_with_cwd_and_envs(test_session):
    """Test command execution with both cwd and envs parameters."""
    cmd = test_session.command
    result = await cmd.execute_command(
        "pwd && echo $CUSTOM_VAR",
        cwd="/tmp",
        envs={"CUSTOM_VAR": "custom_value"}
    )
    
    assert result.success, "Command should succeed"
    assert result.exit_code == 0, "Exit code should be 0"
    assert "/tmp" in result.stdout, "Working directory should be /tmp"
    
    print(f"✓ Combined cwd and envs test passed")
    print(f"  Output: {result.stdout}")


@pytest.mark.asyncio
async def test_command_backward_compatibility(test_session):
    """Test backward compatibility: output field should still work."""
    cmd = test_session.command
    result = await cmd.execute_command("echo 'backward compatible'")
    
    # Verify backward compatibility
    assert hasattr(result, 'output'), "output field should exist for backward compatibility"
    assert result.output is not None, "output should not be None"
    
    # output should be stdout if available, otherwise stderr
    if result.stdout:
        assert result.output == result.stdout, "output should equal stdout when stdout is available"
    elif result.stderr:
        assert result.output == result.stderr, "output should equal stderr when stdout is empty"
    
    print(f"✓ Backward compatibility test passed: output={result.output}")


@pytest.mark.asyncio
async def test_command_path_env(test_session):
    """Test PATH environment variable."""
    cmd = test_session.command
    result = await cmd.execute_command("echo $PATH")
    assert result.success
    assert len(result.output) > 0
    print(f"PATH: {result.output[:100]}...")

File: java/agentbay/src/test/java/com/aliyun/agentbay/test/TestCommandEnvironment.java
package com.aliyun.agentbay.test;

import com.aliyun.agentbay.AgentBay;
import com.aliyun.agentbay.exception.AgentBayException;
import com.aliyun.agentbay.model.CommandResult;
import com.aliyun.agentbay.model.DeleteResult;
import com.aliyun.agentbay.model.SessionResult;
import com.aliyun.agentbay.session.CreateSessionParams;
import com.aliyun.agentbay.session.Session;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.*;

public class TestCommandEnvironment {

    private static AgentBay agentBay;
    private static Session session;

    private static String getTestApiKey() {
        String apiKey = System.getenv("AGENTBAY_API_KEY");
        if (apiKey == null || apiKey.trim().isEmpty()) {
            apiKey = "akm-xxx";
            System.out.println("Warning: Using default API key. Set AGENTBAY_API_KEY environment variable for testing.");
        }
        return apiKey;
    }

    @BeforeClass
    public static void setUp() throws AgentBayException {
        System.out.println("Setting up test environment...");
        String apiKey = getTestApiKey();
        agentBay = new AgentBay();

        CreateSessionParams params = new CreateSessionParams();
        params.setImageId("linux_latest");
        SessionResult sessionResult = agentBay.create(params);

        assertTrue("Failed to create session: " + sessionResult.getErrorMessage(),
                sessionResult.isSuccess());
        assertNotNull("Session object is null", sessionResult.getSession());

        session = sessionResult.getSession();
        System.out.println("✅ Session created with ID: " + session.getSessionId());
    }

    @AfterClass
    public static void tearDown() {
        if (session != null && agentBay != null) {
            try {
                System.out.println("Cleaning up session...");
                DeleteResult deleteResult = agentBay.delete(session, false);
                if (deleteResult.isSuccess()) {
                    System.out.println("✅ Session deleted successfully");
                } else {
                    System.err.println("⚠️ Failed to delete session: " + deleteResult.getErrorMessage());
                }
            } catch (Exception e) {
                System.err.println("⚠️ Error during cleanup: " + e.getMessage());
            }
        }
    }

    @Test
    public void testCommandWithCwd() {
        System.out.println("\n📁 Testing command with cwd parameter...");

        CommandResult result = session.getCommand().executeCommand("pwd", 5000, "/tmp", null);

        assertTrue("Command should succeed", result.isSuccess());
        assertEquals("Exit code should be 0", 0, result.getExitCode());
        assertTrue("Working directory should be /tmp, got: " + result.getOutput(),
                result.getOutput().contains("/tmp"));

        System.out.println("✅ CWD test passed: working directory=" + result.getOutput().trim());
    }

    @Test
    public void testCommandWithEnvs() {
        System.out.println("\n🌍 Testing command with envs parameter...");

        Map<String, String> envs = new HashMap<>();
        envs.put("TEST_VAR", "test_value_123");

        CommandResult result = session.getCommand().executeCommand("echo $TEST_VAR", 5000, null, envs);

        assertTrue("Command should succeed", result.isSuccess());
        assertEquals("Exit code should be 0", 0, result.getExitCode());

        String output = result.getOutput().trim();
        if (output.contains("test_value_123")) {
            System.out.println("✅ Envs test passed: environment variable set correctly: " + output);
        } else {
            System.out.println("⚠️ Envs test: environment variable may not be set (output: " + output + ")");
        }
    }

    @Test
    public void testCommandWithCwdAndEnvs() {
        System.out.println("\n🔀 Testing command with both cwd and envs parameters...");

        Map<String, String> envs = new HashMap<>();
        envs.put("CUSTOM_VAR", "custom_value");

        CommandResult result = session.getCommand().executeCommand(
                "pwd && echo $CUSTOM_VAR",
                5000,
                "/tmp",
                envs
        );

        assertTrue("Command should succeed", result.isSuccess());
        assertEquals("Exit code should be 0", 0, result.getExitCode());
        assertTrue("Working directory should be /tmp", result.getOutput().contains("/tmp"));

        System.out.println("✅ Combined cwd and envs test passed");
        System.out.println("  Output: " + result.getOutput());
    }

    @Test
    public void testCommandWithCwdContainingSpaces() {
        System.out.println("\n📂 Testing command with cwd containing spaces...");

        String testDir = "/tmp/test dir with spaces";

        CommandResult mkdirResult = session.getCommand().executeCommand(
                "mkdir -p '" + testDir + "'",
                5000,
                null,
                null
        );
        assertTrue("Should be able to create directory with spaces", mkdirResult.isSuccess());

        CommandResult result = session.getCommand().executeCommand("pwd", 5000, testDir, null);

        assertTrue("Command should succeed with cwd containing spaces", result.isSuccess());
        assertEquals("Exit code should be 0", 0, result.getExitCode());
        assertTrue("Working directory should contain test dir, got: " + result.getOutput(),
                result.getOutput().contains(testDir) || result.getOutput().contains("/tmp/test"));

        CommandResult cleanupResult = session.getCommand().executeCommand(
                "rm -rf '" + testDir + "'",
                5000,
                null,
                null
        );

        System.out.println("✅ CWD with spaces test passed: directory=" + testDir);
    }

    @Test
    public void testCommandWithEnvsSpecialCharacters() {
        System.out.println("\n🎭 Testing command with environment variables containing special characters...");

        Map<String, String> envs = new HashMap<>();
        envs.put("TEST_VAR", "value with 'single quotes'");

        CommandResult result = session.getCommand().executeCommand("echo $TEST_VAR", 5000, null, envs);

        assertTrue("Command should succeed with env containing single quotes", result.isSuccess());
        assertEquals("Exit code should be 0", 0, result.getExitCode());

        String output = result.getOutput().trim();
        if (output.contains("value with") && output.contains("single quotes")) {
            System.out.println("✅ Envs with single quotes test passed: " + output);
        } else {
            System.out.println("⚠️ Envs with single quotes: output may not match exactly: " + output);
        }
    }

    @Test
    public void testCommandInvalidEnvs() {
        System.out.println("\n❌ Testing command with invalid envs (should throw exception)...");

        Map<String, String> envs = new HashMap<>();
        envs.put("TEST_VAR", "123");

        try {
            CommandResult result = session.getCommand().executeCommand("echo test", 5000, null, envs);
            System.out.println("✅ Valid envs (all strings) test passed");
        } catch (Exception e) {
            fail("Valid envs should not throw exception: " + e.getMessage());
        }
    }

    @Test
    public void testCommandBackwardCompatibility() {
        System.out.println("\n🔄 Testing backward compatibility (no cwd, no envs)...");

        CommandResult result = session.getCommand().executeCommand("echo 'backward compatible'", 5000);

        assertTrue("Command should succeed", result.isSuccess());
        assertNotNull("Output should not be null", result.getOutput());

        System.out.println("✅ Backward compatibility test passed: output=" + result.getOutput());
    }

    @Test
    public void testCommandPathEnv() {
        System.out.println("\n🛤️ Testing PATH environment variable...");

        CommandResult result = session.getCommand().executeCommand("echo $PATH", 5000);

        assertTrue("Command should succeed", result.isSuccess());
        assertTrue("PATH should not be empty", result.getOutput().length() > 0);

        System.out.println("✅ PATH: " + result.getOutput().substring(0, Math.min(100, result.getOutput().length())) + "...");
    }

    @Test
    public void testCommandCwdAndEnvsWithSpecialChars() {
        System.out.println("\n🎪 Testing command with both cwd (spaces) and envs (special chars)...");

        String testDir = "/tmp/test dir with spaces";
        CommandResult mkdirResult = session.getCommand().executeCommand(
                "mkdir -p '" + testDir + "'",
                5000,
                null,
                null
        );
        assertTrue("Should be able to create directory with spaces", mkdirResult.isSuccess());

        Map<String, String> envs = new HashMap<>();
        envs.put("CUSTOM_VAR", "value with 'quotes' and ; semicolon");

        CommandResult result = session.getCommand().executeCommand(
                "pwd && echo $CUSTOM_VAR",
                5000,
                testDir,
                envs
        );

        assertTrue("Command should succeed with both cwd (spaces) and envs (special chars)", result.isSuccess());
        assertEquals("Exit code should be 0", 0, result.getExitCode());
        assertTrue("Working directory should contain test dir",
                result.getOutput().contains(testDir) || result.getOutput().contains("/tmp/test"));

        String output = result.getOutput().trim();
        if (output.contains("value") || output.contains("CUSTOM_VAR")) {
            System.out.println("✅ Combined cwd (spaces) and envs (special chars) test passed");
            System.out.println("  Output: " + output.substring(0, Math.min(100, output.length())) + "...");
        } else {
            System.out.println("⚠️ Combined test: output may not show env var: " + output);
        }

        CommandResult cleanupResult = session.getCommand().executeCommand(
                "rm -rf '" + testDir + "'",
                5000,
                null,
                null
        );
    }

    public static void main(String[] args) {
        System.out.println("=== Running Command Environment Tests ===\n");

        TestCommandEnvironment test = new TestCommandEnvironment();

        try {
            setUp();

            test.testCommandWithCwd();
            test.testCommandWithEnvs();
            test.testCommandWithCwdAndEnvs();
            test.testCommandWithCwdContainingSpaces();
            test.testCommandWithEnvsSpecialCharacters();
            test.testCommandInvalidEnvs();
            test.testCommandBackwardCompatibility();
            test.testCommandPathEnv();
            test.testCommandCwdAndEnvsWithSpecialChars();

            tearDown();

            System.out.println("\n=== ✅ All Command Environment Tests Completed Successfully ===");
        } catch (Exception e) {
            System.err.println("\n=== ❌ Test Failed ===");
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
}
